google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -402,30 +417,53 @@ private Session take() throws SpannerException {
     int numSessionsToClose = 0;
     int sessionsToClosePerLoop = 0;
 
+    @GuardedBy("lock")
+    ScheduledFuture<?> scheduledFuture;
+
+    @GuardedBy("lock")
+    boolean running;
+
     void init() {
       // Scheduled pool maintenance worker.
-      executor.scheduleAtFixedRate(
-          new Runnable() {
-            @Override
-            public void run() {
-              maintainPool();
-            }
-          },
-          LOOP_FREQUENCY,
-          LOOP_FREQUENCY,
-          TimeUnit.MILLISECONDS);
+      scheduledFuture =
+          executor.scheduleAtFixedRate(
+              new Runnable() {
+                @Override
+                public void run() {
+                  maintainPool();
+                }
+              },
+              LOOP_FREQUENCY,
+              LOOP_FREQUENCY,
+              TimeUnit.MILLISECONDS);
+    }
+
+    void close() {
+      scheduledFuture.cancel(false);
+      if (!running) {
```

Comment:
The reason maintainPool needs to manage lock on its own is because we cannot hold locks for the entire duration of that method. But I went ahead and added locks to close and init as well just to be sure.